mass_mailing_partner/models/mailing_contact.py

```diff
@@ -42,16 +42,20 @@ def _check_partner_id_list_ids(self):
                         + ": %s" % contact.partner_id.display_name
                     )
 
+    def update_fields_based_on_partner(self):
+        for rec in self:
+            if rec.partner_id:
+                rec.name = rec.partner_id.name
+                rec.email = rec.partner_id.email
+                rec.title_id = rec.partner_id.title
+                rec.company_name = (
+                    rec.partner_id.company_id.name or rec.partner_id.company_name
+                )
+                rec.country_id = rec.partner_id.country_id
```

Comment:
To reduce one complexity level:

```suggestion
        for rec in self.filtered("partner_id"):
              rec.name = rec.partner_id.name
              rec.email = rec.partner_id.email
              rec.title_id = rec.partner_id.title
              rec.company_name = (
                  rec.partner_id.company_id.name or rec.partner_id.company_name
              )
              rec.country_id = rec.partner_id.country_id
```